broker-core/src/test/java/io/zeebe/broker/workflow/variables/WorkflowInstanceVariableTest.java

```diff
@@ -0,0 +1,237 @@
+
+  @Before
+  public void init() {
+    testClient = apiRule.partitionClient();
+  }
+
+  @Test
+  public void shouldCreateVariableByWorkflowInstanceCreation() {
+    // given
+    testClient.deploy(WORKFLOW);
+
+    // when
+    final long workflowInstanceKey = testClient.createWorkflowInstance(PROCESS_ID, "{'x':1}");
+
+    // then
+    final Record<VariableRecordValue> variableRecord =
+        RecordingExporter.variableRecords(VariableIntent.CREATED).getFirst();
+    Assertions.assertThat(variableRecord.getValue())
+        .hasScopeInstanceKey(workflowInstanceKey)
+        .hasName("x")
+        .hasValue("1");
+  }
+
+  @Test
+  public void shouldCreateVariableByJobCompletion() {
+    // given
+    testClient.deploy(WORKFLOW);
+
+    final long workflowInstanceKey = testClient.createWorkflowInstance(PROCESS_ID);
+
+    // when
+    testClient.completeJobOfType("test", "{'x':1}");
+
+    // then
+    // TODO (saig0) #1860: currently the job payload is always set in the scope of the task
+    final Record<VariableRecordValue> variableRecord =
+        RecordingExporter.variableRecords(VariableIntent.CREATED)
+            .withScopeInstanceKey(workflowInstanceKey)
+            .getFirst();
+    Assertions.assertThat(variableRecord.getValue())
+        .hasScopeInstanceKey(workflowInstanceKey)
+        .hasName("x")
+        .hasValue("1");
+  }
+
+  @Test
+  public void shouldCreateVariableByOutputMapping() {
+    // given
+    testClient.deploy(
+        Bpmn.createExecutableProcess(PROCESS_ID)
+            .startEvent()
+            .serviceTask("task", t -> t.zeebeTaskType("test").zeebeOutput("$.x", "$.y"))
+            .endEvent()
+            .done());
+
+    final long workflowInstanceKey = testClient.createWorkflowInstance(PROCESS_ID);
+
+    // when
+    testClient.completeJobOfType("test", "{'x':1}");
+
+    // then
+    final Record<VariableRecordValue> variableRecord =
+        RecordingExporter.variableRecords(VariableIntent.CREATED).withName("y").getFirst();
+    Assertions.assertThat(variableRecord.getValue())
+        .hasScopeInstanceKey(workflowInstanceKey)
+        .hasName("y")
+        .hasValue("1");
+  }
+
+  @Test
+  public void shouldCreateVariableByUpdatePayload() {
+    // given
+    testClient.deploy(WORKFLOW);
+
+    final long workflowInstanceKey = testClient.createWorkflowInstance(PROCESS_ID);
+
+    // when
+    testClient.updatePayload(workflowInstanceKey, "{'x':1}");
+
+    // then
+    final Record<VariableRecordValue> variableRecord =
+        RecordingExporter.variableRecords(VariableIntent.CREATED).getFirst();
+    Assertions.assertThat(variableRecord.getValue())
+        .hasScopeInstanceKey(workflowInstanceKey)
+        .hasName("x")
+        .hasValue("1");
+  }
+
+  @Test
+  public void shouldCreateMultipleVariablesFromPayload() {
+    // given
+    testClient.deploy(WORKFLOW);
+
+    // when
+    testClient.createWorkflowInstance(PROCESS_ID, "{'x':1, 'y':2}");
```

Comment:
Can we add a test for updating an existing varaible and creating a new. So one variable already exist and the other is newly created

Reply:
Sure, I'll create a new case for this.